llama_index/vector_stores/pinecone.py

```diff
@@ -243,6 +251,43 @@ def __init__(
             batch_size=batch_size,
             remove_text_from_metadata=remove_text_from_metadata,
         )
+        pinecone = _import_pinecone()
+
+        # Cannot pass in str for pinecone_index
+        # Both clients >= 3.0.0 and < 3.0.0 should have pinecone.Index class
+        if not isinstance(pinecone_index, pinecone.Index):
+            raise ValueError(
+                f"`pinecone_index` should be an instance of pinecone.Index, "
+                f"got {type(pinecone_index)}"
+            )
+
+        self._pinecone_index = pinecone_index or self._initialize_pinecone_client(
+            api_key, index_name, environment, **kwargs
+        )
+
+    @staticmethod
```

Comment:
We should probably document at least this function pretty extensively for future maintainers and explain that it's an adapter pattern that is encapsulating our dual stack version logic temporarily, until the period after which we're no longer going to support multiple Pinecone client versions.